src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,194 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Net.Http;
+using System.Threading.Channels;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Controls the WebTransport session and keeps track of all the streams.
+/// </summary>
+internal class WebTransportSession : IWebTransportSession
+{
+    private readonly ConcurrentDictionary<long, WebTransportStream> _openStreams = new();
+    private readonly Channel<WebTransportStream> _pendingStreams;
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _connectStream = default!;
+    private bool _isClosing;
+
+    private static readonly ReadOnlyMemory<byte> OutputStreamHeader = new(new byte[] {
+            0x40 /*quic variable-length integer length*/,
+            (byte)Http3StreamType.WebTransportUnidirectional,
+            0x00 /*body*/});
+
+    internal static string WebTransportProtocolValue => "webtransport";
+    internal static string SecPrefix => "sec-webtransport-http3-";
+    internal static string VersionHeaderPrefix => $"{SecPrefix}draft";
+
+    // Order is important as we choose the first supported
+    // version (preferenc dereases as index increases)
+    internal static readonly IEnumerable<string> SuppportedWebTransportVersions = new string[]
+    {
+        $"{VersionHeaderPrefix}02"
+    };
+
+    long IWebTransportSession.SessionId => _connectStream.StreamId;
+
+    internal WebTransportSession(Http3Connection connection, Http3Stream connectStream)
+    {
+        _connection = connection;
+        _connectStream = connectStream;
+        _isClosing = false;
+        _pendingStreams = Channel.CreateBounded<WebTransportStream>(100); // todo what should the capacity be?
```

Comment:
Should this limit be configurable by the application (i.e. a KestrelServerOption or a parameter to the constructor) or should I just hardcode it? If the latter, what would be a good number?

Reply:
For context, there is this discussion going on in the spec repo that may or may not eventually influence this number: https://github.com/ietf-wg-webtrans/draft-ietf-webtrans-http3/pull/74

Reply:
After an offline discussion with Chris, we decided that this isn't a good place to bound the streams. Thus, making it unbounded here.